A14-channelz.md

```diff
@@ -0,0 +1,1051 @@
+such as connecting clients, timeouts, protocol violations, etc.  It also
+includes information about the shutdown of the server.  This is important as the
+server may have many active calls when it is shutting down, which prevent it
+from terminating.
+
+### Sockets and Listen Sockets
+
+Unlike channels, servers have two kinds of sockets.  A regular socket is one
+that is connected to a remote endpoint.  This is the most common type of socket.
+Servers also have a "listen" socket, which is used to accept and create new
+sockets.
+
+There may be no listen sockets, but active regular sockets.  This can occur
+during shutdown, when the server no longer wishes to accept connections but
+would like to gracefully complete the existing ones.  If a socket could exist in
+both locations, it may.  Implementations should prefer to segregate the two
+sockets.
+
+
+## Sockets
+
+Conceptually, sockets are the equivalent of a file descriptor.  They have a
+local and remote address, as well as some concept of security detail.  Sockets
+keep track of "streams" while Channels and Servers keep track of "calls."
+Sockets also expose access to underlying details of the operating system, such
+as socket options.  They are the lowest level in the channelz hierarchy.
+
+A SocketRef is associated with each socket, and contains a unique id and an
+optional name.  The id is unique among even the channels and servers.  The name
+includes information about the socket in a human readable form.  This is
+important as it is used as a "bread crumb" when deciding which socket to query
+for, from the context of a channel or server.
+
+### Socket Data
+
+The data exposed by a socket is much lower level than the channel or server
+data.  This information is meant to aid in debugging slow or stalled
+connections.  By default it includes the following:
+
+* Number of streams started, succeeded, or failed.
+* Number of gRPC messages sent and received.
+* Number of keepalive messages sent.
+* The last time a outbound stream was started and the last time an inbound
+  stream was started.
+* The last time a outbound message was sent and the last time an inbound
+  message was received.
+* The amount of flow control window granted to the remote endpoint, and the
+  amount of flow control window granted by remote endpoint.
+* All known socket options set.
+
+The stream counts differ slightly than that of the number of calls.  Calls may
+be started before a socket (transport) is available, so the stream counter may
+smaller than the corresponding call counters.  Conversely, if a stream fails and
+the call is retriable, a new stream may be started.  This could make the stream
+counts higher than the call counts.  Streams are considered successful if they
+have the HTTP/2 EoS bit set, or failed if terminated otherwise (usually an
+HTTP/2 RST_STREAM frame).
+
+The message counts represent how many gRPC messages are sent and received.
```

Comment:
It seems a little inconsistent for the socket to track streams instead of calls (which implies that it's working at the HTTP/2 layer, not the gRPC layer) but then track a gRPC-layer entity like messages.  Maybe the socket should track both streams and calls?  Or maybe it shouldn't track messages at all?

Reply:
I'm not sure why bytes sent is hard to track.  But even if they are, I think it would make more sense for the implementation to provide an estimate of bytes sent instead of breaking the abstraction between layers this way.

I am really not comfortable with mixing the two layers this way.  If you do not think there's an alternative, then we should get @a11r to weigh in on this.

Reply:
Bytes is hard because it means different things to different people.  For each of the following questions, there are reasonable people who would disagree about if they count:

* Http Proxy overhead
* Http/2 ping frames
* Http/2 window frames
* Http/2 padding
* Http2 Hpack / continuation frames
* Per message compression and / or whole stream compression
* SSL Framing overhead (and does this include overhead added by hardware crypto?)
* TCP framing overhead
* TCP keep alives
* TCP sidechannel
* TCP retransmissions and duplicate acks
* Quic overhead

Secondly, not all transports report all of these numbers, like Cronet, Inprocess, and possibly gRPC over gRPC.

Thirdly, we don't always know when something is "sent".   Does it get incremented at the beginning of a 1gig message or at the beginning, or update it as it sends.  

Your point about abstraction layer is valid; I don't think there is an easy way to update the upper layers of gRPC from the lower level transport.   I picked the stream because at this point it is certain which transport is servicing the message.   

Like I said, this is just a proxy for something that is easy to calculate.  It doesn't solve all the problems mentioned above, but it is a useful metric.   We can talk about adding more things based on user feedback in 1.1.